src/main/java/net/minecraftforge/event/terraingen/DecorateBiomeEvent.java

```diff
@@ -96,27 +98,94 @@ public Post(World world, Random rand, BlockPos pos)
 
     /**
      * This event is fired when a chunk is decorated with a biome feature.
-     *
+     * The amount denotes how many instances of the feature will be generated.
+     * <p>
      * You can set the result to DENY to prevent the default biome decoration.
      */
     @HasResult
     public static class Decorate extends DecorateBiomeEvent
     {
+        private final EventType type;
+        @Nullable
+        private final Integer totalAmount;
+        @Nullable
+        private Integer modifiedAmount;
+
+        @Deprecated
+        public Decorate(World world, Random rand, BlockPos pos, EventType type)
+        {
+            super(world, rand, pos);
+            this.type = type;
+            this.totalAmount = null;
+            this.modifiedAmount = null;
+        }
+
+        public Decorate(World world, Random rand, BlockPos pos, EventType type, int amount)
+        {
+            super(world, rand, pos);
+            this.type = type;
+            this.totalAmount = amount;
+            this.modifiedAmount = amount;
+        }
+
         public EventType getType()
         {
             return type;
         }
 
-        /** Use CUSTOM to filter custom event types
-         */
-        public static enum EventType { BIG_SHROOM, CACTUS, CLAY, DEAD_BUSH, DESERT_WELL, LILYPAD, FLOWERS, FOSSIL, GRASS, ICE, LAKE_WATER, LAKE_LAVA, PUMPKIN, REED, ROCK, SAND, SAND_PASS2, SHROOM, TREE, CUSTOM }
+        public boolean hasAmountData()
+        {
+            return totalAmount != null && modifiedAmount != null;
+        }
 
-        private final EventType type;
+        public int getTotalAmount()
+        {
+            if (totalAmount == null)
+                throw new IllegalStateException();
 
-        public Decorate(World world, Random rand, BlockPos pos, EventType type)
+            return totalAmount;
+        }
+
+        public int getModifiedAmount()
         {
-            super(world, rand, pos);
-            this.type = type;
+            if (modifiedAmount == null)
+                throw new IllegalStateException();
+
+            return modifiedAmount;
+        }
+
+        public void setModifiedAmount(int modifiedAmount)
+        {
+            this.modifiedAmount = modifiedAmount;
+        }
+
+        /**
+         * Use CUSTOM to filter custom event types
+         */
+        public enum EventType
+        {
+            BIG_SHROOM, CACTUS, CLAY, DEAD_BUSH, DESERT_WELL, LILYPAD, FLOWERS, FOSSIL, GRASS, ICE, LAKE_WATER, LAKE_LAVA, PUMPKIN, REED, ROCK, SAND, SAND_PASS2, SHROOM, TREE, CUSTOM
+        }
+
+        public static class Generator extends Decorate
+        {
+            public final WorldGenerator generator;
+            public final Position position;
+
+            public enum Position
+            {
```

Comment:
My thinking was to make this parameter a bit more generally useful. A generator usually is passed in just the constructor parameters (which a mod can read back out of the generator field) and the position where to generate - which is unaccessible in this event otherwise. It's the only thing distinguishing surface and cave mushrooms and thus covers that case fine, and while vanilla doesn't otherwise need this field it will probably be useful in modded cases a bit more.